dingdongg/spotify-payment-backend
# Create login feature

## Description
- Implemented login feature and related security features
    - uses Redis as session storage
- Improve error logging
- refactor different modules into different routes

*This is a big PR so it can be kinda tricky to test. Let me know if you run into any trouble and/or to hop on a call to walk through it together*

## Testing

- before testing, run `npm install`
- you also have to install redis (see [this link](https://redis.io/docs/getting-started/installation/))
- if you want to see the state of the session storage throughout testing, open a new terminal and type `redis-cli`. 
    - the following are useful commands:
        - `keys *`: gets all sessions stored in memory at that moment
        - `get <session_id>`: retrieve a particular session
        - `flushall`: empties the session storage

__Flow #1:__
- try calling an endpoint that would typically require the user to be logged in
      - should get back 401 error, saying the user isn't authenticated
- call `GET /csrf-token` (this is the token to use whenever you make state-changing requests within this session)
- call `POST /login`, using a `x-csrf-token` header with value from the previous step
- user should now be authenticated (logged in)
- try calling an endpoint that would typically require the user to be logged in 
      - this time, the API call should be successful
- try calling `DELETE /users/:user_id` (create a test user to delete if you want)
      - if you are logged in to an admin account, the user should be deleted
      - otherwise, you should get back 403 error, saying the user isn't authorized to access this resource

__Flow #2:__
- call `GET /csrf-token`
- call `POST /login`, using a `x-csrf-token` header with value from the previous step
- call `POST /logout`, again using the CSRF token from step 1
- should be successfully logged out 
- in Redis, the login session should be destroyed
